File: backend/src/jobs/subscriptionExpiryJob.js
// src/jobs/subscriptionExpiryJob.js
// Corre diariamente: avisa Company Admin + Financeiro em patamares (30/7/3/1/0
// dias e durante o período de tolerância) antes/depois de a subscrição vencer.
// Mesmo padrão de policyExpiryJob.js — a lógica em si vive em
// assinaturaService.enviarAvisosDeExpiracao, este ficheiro só agenda.

const cron = require('node-cron');
const logger = require('../config/logger');
const assinaturaService = require('../services/assinaturaService');
const alertaOperacional = require('../services/alertaOperacionalService');

function scheduleSubscriptionExpiryJob() {
  // Todos os dias às 07:00 — mesma hora do aviso de expiração de apólices.
  cron.schedule('0 7 * * *', async () => {
    try {
      const count = await assinaturaService.enviarAvisosDeExpiracao();
      if (count > 0) {
        logger.info(`Avisos de expiração de subscrição enviados: ${count}`);
      }
    } catch (err) {
      logger.error('Falha ao processar avisos de expiração de subscrição', { message: err.message });
      await alertaOperacional.avisarFalha(
        'EXPIRACAO_SUBSCRICOES',
        'os avisos de expiração de subscrição falharam',
        `O processamento não correu.\n\nErro: ${err.message}\n\n`
        + 'Uma subscrição que vence sem aviso é receita perdida em silêncio — a empresa não sabe '
        + 'que tem de renovar, e a KIXIMA só descobre quando alguém olhar para a fila do Admin.',
      ).catch(() => {});
    }
  });
}

module.exports = { scheduleSubscriptionExpiryJob };

File: backend-java/src/main/java/ao/kixima/plan/SubscriptionExpiryJob.java
package ao.kixima.plan;

import ao.kixima.ops.OperationalAlertService;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.boot.autoconfigure.condition.ConditionalOnProperty;
import org.springframework.scheduling.annotation.Scheduled;
import org.springframework.stereotype.Component;

/**
 * Espelha backend/src/jobs/subscriptionExpiryJob.js — mesmo padrão do
 * PolicyExpiryJob (cron diário às 07:00, mesma hora do aviso de apólices).
 *
 * DESLIGADO POR OMISSÃO ({@code kixima.jobs.expiracao-subscricoes.enabled: false})
 * — o Node continua a correr o seu próprio; ligar os dois duplicaria o aviso.
 */
@Component
@ConditionalOnProperty(name = "kixima.jobs.expiracao-subscricoes.enabled", havingValue = "true")
public class SubscriptionExpiryJob {

    private static final Logger log = LoggerFactory.getLogger(SubscriptionExpiryJob.class);

    private final SubscriptionExpiryService subscriptionExpiryService;
    private final OperationalAlertService operationalAlertService;

    public SubscriptionExpiryJob(SubscriptionExpiryService subscriptionExpiryService,
                                  OperationalAlertService operationalAlertService) {
        this.subscriptionExpiryService = subscriptionExpiryService;
        this.operationalAlertService = operationalAlertService;
    }

    @Scheduled(cron = "0 0 7 * * *")
    public void correr() {
        try {
            subscriptionExpiryService.enviarAvisosDeExpiracao();
        } catch (Exception e) {
            log.error("Falha ao processar avisos de expiração de subscrição: {}", e.getMessage());
            operationalAlertService.avisarFalha("EXPIRACAO_SUBSCRICOES", "os avisos de expiração de subscrição falharam",
                    "O processamento não correu.\n\nErro: " + e.getMessage() + "\n\n"
                            + "Uma subscrição que vence sem aviso é receita perdida em silêncio — a empresa não sabe "
                            + "que tem de renovar, e a KIXIMA só descobre quando alguém olhar para a fila do Admin.");
        }
    }
}
